Jetpack CP: Jetpack install vm and repository.

hafizrahman opened the pull request:
## Description:

This is the second part for the "Jetpack CP: Install Jetpack from app" feature: https://github.com/woocommerce/woocommerce-android/issues/4865#issuecomment-956128458

The PR adds: 
- a `PluginRepository` with functionalities to install and fetch plugin
- a `JetpackInstallViewModel` with Jetpack install functionality and install progress handling in its viewstate
- Unit test for `JetpackInstallViewModel`

## To test:

Run `JetpackInstallViewModelTest` and make sure all tests succeed.

hafizrahman committed:
Update: Use latest FluxC changes.

hafizrahman committed:
Add: Jetpack install repository.

hafizrahman committed:
Add: Fetch plugin and parcelized Plugin Status in repository.

hafizrahman committed:
Add: activate plugin in repository, as well as some smaller tweaks.

hafizrahman committed:
Add: Jetpack install view model. Removes unneeded values in repository.

hafizrahman committed:
Refactor: Make repository be more generic for any plugins.

hafizrahman committed:
Merge branch 'develop' into hafiz/jetpack-install-2

# Conflicts:
#	build.gradle

hafizrahman committed:
Refactor: Move view model to jetpack package.

hafizrahman committed:
Merge branch 'develop' into hafiz/jetpack-install-2

# Conflicts:
#	build.gradle

hafizrahman committed:
Refactor: fetchJetpackSitePlugin() to be suspend function.

hafizrahman committed:
Merge branch 'develop' into hafiz/jetpack-install-2

hafizrahman committed:
Refactor: Simplify install flow, without retry mechanism.

hafizrahman committed:
Refactor: Use coroutines for simulating connecting/finish steps.

hafizrahman committed:
Merge branch 'develop' into hafiz/jetpack-install-2

hafizrahman committed:
Fix: Remove test plugin.

hafizrahman committed:
Add: Jetpack Install VM unit tests.

hafizrahman committed:
Update: sync build.gradle to match current develop.

For some reason this branch's version was still outdated even after merging latest develop to it.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/PluginRepository.kt:
```diff
@@ -0,0 +1,139 @@
+package com.woocommerce.android.ui.mystore
```
❓ Any suggestion about what would be a good place for this instead of `/ui/mystore` ? I looked at existing packages and didn't see any particularly good place for it.

JorgeMucientes replied:
I have the same doubts too! Regarding where to place repositories. Because it doesn't feel right to place repositories inside a ui package. 
I guess we should open a discussion with the team on how to structure this. We could move the feature package to the top and then inside separate ui and data packes. Something like this:

```
├───mystore
│   ├───data
│   │    PluginRepository.kt
│   ├───ui
│   │    JetpackInstallViewModel.kt
│   │    Fragment.kt ...

...
```

hichamboushaba replied:
>  guess we should open a discussion with the team on how to structure this
+1

For this PR, I'd say move it to the `jetpack` package since it's there where it's used right now, and as it doesn't have any relation with `mystore` feature.
After the discussion with the team, we can decide how to refactor the packaging for the whole app.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModel.kt:
```diff
@@ -0,0 +1,94 @@
+        object Finished : InstallStatus()
+
+        @Parcelize
+        data class Failed(val error: String?) : InstallStatus()
```
Really minor, but this `error: String?` shouldn't be nullable right? There is always going to be an error message, at least a generic one right?

JorgeMucientes commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModelTest.kt:
```diff
@@ -0,0 +1,112 @@
+
+    @Before
+    fun setup() {
+        viewModel = spy(
```
I'm curious why a Spy is needed here? You are not tracking/verifying interactions over the view model itself, nor mocking any of it's methods right ?

hafizrahman replied:
Thanks for catching this! I was testing things around and forgot to change it.

JorgeMucientes commented:
Nice and clean @hafizrahman, well done! Just added minor code change suggestions :) You are welcomed to apply them or not.

hafizrahman committed:
Refactor: Move PluginRepository to be under ui/jetpack for now.

More details on this conversation: https://github.com/woocommerce/woocommerce-android/pull/5287#pullrequestreview-810997145

hafizrahman committed:
Refactor: Failed InstallStatus does not need to be nullable.

hafizrahman committed:
Fix: Remove usage of spy in test.

hafizrahman commented:
Thank you for your suggestions @JorgeMucientes @hichamboushaba 

I've addressed them and will merge now :)
